Generic types in structs cause C error

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** V 0.2.2 298fc53.a6764e3
**OS:** linux, "Arch Linux"

**What did you do?**
```v
struct Node<T> {
mut:
	val T
	next &Node<T>
}

fn make_node<T>(val []T) Node<T> {
	return Node{
		val: val[0]
		next: 0
	}
}

fn main() {
	n := make_node([1, 2, 3])
	println(n.val)
}
```
`v run .`

**What did you expect to see?**
successful compile, 1 being written to stdout
 
**What did you see instead?**
```
==================
/tmp/v/generics.5634080792445069707.tmp.c:874: error: ';' expected (got "main__Node_T_T")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
```
the C lines:
```
 872 struct main__Node_T_int {
 873         int val;
 874         main__Node_T_T* next;    
 875 } ;
 ```
 I assume that the bug is that `main__Node_T_T` is not replaced by `main__Node_T_int`. (`main__Node_T_T` is undefined)

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1618,7 +1618,7 @@ fn (mut c Checker) check_return_generics_struct(return_type ast.Type, mut call_e
 				mut fields := rts.info.fields.clone()
 				if rts.info.generic_types.len == concrete_types.len {
 					generic_names := rts.info.generic_types.map(c.table.get_type_symbol(it).name)
-					for i, _ in fields {
+					for i in 0 .. fields.len {
 						if t_typ := c.table.resolve_generic_to_concrete(fields[i].typ,
 							generic_names, concrete_types, true)
 						{

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1620,7 +1620,7 @@ fn (mut c Checker) check_return_generics_struct(return_type ast.Type, mut call_e
 					generic_names := rts.info.generic_types.map(c.table.get_type_symbol(it).name)
 					for i, _ in fields {
 						if t_typ := c.table.resolve_generic_to_concrete(fields[i].typ,
-							generic_names, concrete_types, false)
+							generic_names, concrete_types, true)
 						{
 							fields[i].typ = t_typ
 						}

--- a/vlib/v/tests/generics_return_recursive_generics_struct_test.v
+++ b/vlib/v/tests/generics_return_recursive_generics_struct_test.v
@@ -1,0 +1,18 @@
+struct Node<T> {
+mut:
+	val  T
+	next &Node<T>
+}
+
+fn make_node<T>(val []T) Node<T> {
+	return Node{
+		val: val[0]
+		next: 0
+	}
+}
+
+fn test_generics_return_recursive_generics_struct() {
+	n := make_node([1, 2, 3])
+	println(n.val)
+	assert n.val == 1
+}

